### ISSUE TITLE
Issue Oracle BulkMerge  : ORA-01400 Cannot insert NULL into

### ISSUE THREAD

**Issue by tuyndv** [zzzprojects/EntityFramework-Extensions#496]
Dear Team,

I using BulkMerge have issue ORA-01400: cannot insert NULL into when have item **INSERT** + **UPDATE**.

```js
//Note test
+ BulkInsert :  Passed  
+ BulkUpdate :  Passed  
+ BulkDelete :  Passed  
+ BulkMerge only items insert : Passed  
+ BulkMerge only items update : Passed  
+ BulkMerge with items (insert and update) : Failed  (table option 1 + 2)
+ BulkMerge with items (insert and update) : Passed  (table option  3 identity)
```

Please check for me. 
Thanks

This my code : 

```sql
-- Table Option 1
CREATE TABLE MST_XXX
(
    ITEM_ID                   NUMBER        not null,
    ITEM_CODE                 VARCHAR2(40)  not null
)

-- Table Option 2
CREATE TABLE MST_XXX
(
    ITEM_ID                   NUMBER        not null,
    ITEM_CODE                 VARCHAR2(40)  not null,
    PRIMARY KEY (ITEM_ID)
)
-- Table Option 3 (IDENTITY)
CREATE TABLE MST_XXX
(
    ITEM_ID NUMBER GENERATED BY DEFAULT AS IDENTITY,
    ITEM_CODE                 VARCHAR2(40)  not null,
    PRIMARY KEY (ITEM_ID)
)
```
```csharp
//Class Mode
[Table("MST_XXX")]
public class MST_XXX 
{
    public int ITEM_ID { get; set; }
    public string ITEM_CODE { get; set; }
}
```

```csharp
// OnModelCreating
modelBuilder.Entity<MST_XXX>(t =>
{
    t.HasKey(c => c.ITEM_ID);
});
```
```csharp
// OnModelCreating (IDENTITY)
modelBuilder.Entity<MST_XXX>(t =>
{
    t.Property(c => c.ITEM_ID).ValueGeneratedOnAdd();
    t.HasKey(c => c.ITEM_ID);
});
```

```csharp
// OracleService
List<MST_XXX> MST_XXXs = new List<MST_XXX>();
MST_XXXs.Add(new MST_XXX()
{
    ITEM_ID = 1,
    ITEM_CODE = "1"

});
MST_XXXs.Add(new MST_XXX()
{
    ITEM_ID = 2,
    ITEM_CODE = "2"
});
MST_XXXs.Add(new MST_XXX()
{
    ITEM_ID = 3,
    ITEM_CODE = "3"
});

using (var trans = _context.Database.BeginTransaction())
{
    try
    {
        await _context.BulkMergeAsync(MST_XXXs, p =>
        {
            p.AutoMapOutputDirection = true;
            p.InsertIfNotExists = true;
            p.AllowDuplicateKeys = true;
        });
        trans.Commit();
        return true;
    }
    catch (Exception ex)
    {
        Log.Error(ex, ex.Message);
        trans.Rollback();
        throw;
    }
}


```
 
### Exception
```
Exception message: ORA-01400: cannot insert NULL into ("SYSTEM"."MST_XXX"."ITEM_ID")
Stack trace:
      at OracleInternal.ServiceObjects.OracleConnectionImpl.VerifyExecution(Int32& cursorId, Boolean bThrowArrayBindRelatedErrors, SqlStatementType sqlStatementType, Int32 arrayBindCount, OracleException& exceptionForArrayBindDML, Boolean& hasMoreRowsInDB, Boolean bFirstIterationDone)
   at OracleInternal.ServiceObjects.OracleCommandImpl.ExecuteNonQuery(String commandText, OracleParameterCollection paramColl, CommandType commandType, OracleConnectionImpl connectionImpl, Int32 longFetchSize, Int64 clientInitialLOBFS, OracleDependencyImpl orclDependencyImpl, Int64[]& scnFromExecution, OracleParameterCollection& bindByPositionParamColl, Boolean& bBindParamPresent, OracleException& exceptionForArrayBindDML, OracleConnection connection, Boolean isFromEF)
   at Oracle.ManagedDataAccess.Client.OracleCommand.ExecuteNonQuery()
   at .(DbCommand , BulkOperation , Int32 )
   at .( , DbCommand )
   at .Execute(List`1 actions)
   at Z.BulkOperations.BulkOperation.Execute()
   at .BulkMerge[T](DbContext this, IEntityType entityType, IEnumerable`1 list, Action`1 options, SavingSelector savingSelector, Boolean forceSpecificTypeMapping)
   at .BulkMerge[T](DbContext this, IEnumerable`1 entities, Action`1 options, Boolean isBulkSaveChanges)
   at DbContextExtensions.BulkMerge[T](DbContext this, IEnumerable`1 entities, Action`1 options)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at DbContextExtensions.BulkMergeAsync[T](DbContext this, IEnumerable`1 entities, Action`1 options, CancellationToken cancellationToken)
   at Sep.Oracle.Module.Services.OracleService.TestConnectAsync() in E:\Sep.Oracle.Module\Services\OracleService.cs:line 94
```

### Further technical details
- EF version: Microsoft.EntityFrameworkCore 6.0.7 +  Oracle.EntityFrameworkCore 6.21.61/Devart.Data.Oracle 10.0.0
- EF Extensions version: Z.EntityFramework.Extensions.EFCore 6.14.3 / 6.14.2
- Database Provider: Oracle Version 19.14.0.0.0 / 21.3.0.0.0

**Comment by tuyndv** [comment]
Dear Team,

I using Z.Dapper.Plus 4.0.31 everything works fine. 

Please check for Z.EntityFramework.Extensions.EFCore

Thanks

```
using (var connection = _context.Database.GetDbConnection())
                    {
                        connection.UseBulkOptions(p =>
                        {
                            p.InsertIfNotExists = true;
                            p.ColumnPrimaryKeyNames = new List<string> { "ITEM_ID" };
                        })
                        .BulkMerge(MST_XXXs);
                        trans.Commit();
                    }
```

**Comment by JonathanMagnan** [comment]
Hello @tuyndv ,

Thank you for reporting, we will look at it.

Best Regards,

Jon

**Comment by JonathanMagnan** [comment]
Hello @tuyndv ,

My developer has looked at this issue.

What happens is that your `ITEM_ID ` is set as an identity value in your model (for all 3 options), but this is not an identity value in your database (for options 1 and 2). So your model and database don't match for these both options.

You currently don't get the error with `SaveChanges` or `Dapper Plus`  because of how it works. But the `BulkInsert`/`BulkMerge` method follows the model more strictly and then acts like always as identity in this case. You can easily bypass this limitation by using the `MergeKeepIdentity` options, but by default, our library keeps following the model.

Option 3 works as the identity value are correctly generated.

To summarize, this is currently an expected behavior if your model doesn't match the database. Unlike `SaveChanges`, the `BulkInsert`/`BulkMerge` method doesn't insert a value even if specified if the column is an identity in the model. You can use `InsertKeepIdentity` or `MergeKeepIdentity` to force this case.

Let me know if the cause and solution are clear enough.

Best Regards,

Jon

**Comment by tuyndv** [comment]
Thanks @JonathanMagnan , I will using MergeKeepIdentity  for this case.
